[CI] SharedBlobCacheWarmingServiceIT testCacheIsWarmedBeforeSearchShardRecovery failing

**Build Scans:**
- [elasticsearch-periodic #12227 / release-tests](https://gradle-enterprise.elastic.co/s/zsezjkcx5rsgo)
- [elasticsearch-intake #39026 / part2](https://gradle-enterprise.elastic.co/s/vdit3hrsqzwoi)
- [elasticsearch-pull-request #142542 / part-2](https://gradle-enterprise.elastic.co/s/2niezpkgnkvmo)
- [elasticsearch-pull-request #142542 / part-2](https://gradle-enterprise.elastic.co/s/hqlr5ivnqa4hi)
- [elasticsearch-pull-request #142500 / part-2](https://gradle-enterprise.elastic.co/s/tecjl2efq2cli)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:stateless:internalClusterTest" --tests "org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT.testCacheIsWarmedBeforeSearchShardRecovery" -Dtests.seed=6982AA156F4DC1C3 -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dlicense.key=x-pack/license-tools/src/test/resources/public.key -Dtests.locale=su-ID -Dtests.timezone=America/Lima -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testCacheIsWarmedBeforeSearchShardRecovery'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.Exception: Test abandoned because suite timeout was reached.
```

**Issue Reasons:**
- [main] 5 failures in test testCacheIsWarmedBeforeSearchShardRecovery (2.2% fail rate in 225 executions)
- [main] 3 failures in step part-2 (2.3% fail rate in 128 executions)
- [main] 2 failures in pipeline elasticsearch-pull-request (1.6% fail rate in 128 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

I made a small amount of progress here. For ease of reference, these are the logs I've been working off:

[vdit3hrsqzwoi-org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT-testCacheIsWarmedBeforeSearchShardRecovery-1-output.txt](https://github.com/user-attachments/files/27722079/vdit3hrsqzwoi-org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT-testCacheIsWarmedBeforeSearchShardRecovery-1-output.txt)

The timeout is while the test is in the `ensureSearchHits(indexName, totalDocs)` at the very end of the test. But a little bit earlier we see this:

```
WARNING: Uncaught exception in thread: Thread[#2374,elasticsearch[node_t3][search][T#4],5,TGRP-SharedBlobCacheWarmingServiceIT]
java.lang.AssertionError: should not have sent a request for VBCC data to the indexing node but sent request GetVirtualBatchedCompoundCommitChunkRequest[[lugyiqbk][0],1,5,229376,16384,-9dkDf3WQQikF_noZnCyCw]
	at __randomizedtesting.SeedInfo.seed([636BFED6FD3A01CD]:0)
	at org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT.lambda$failObjectStoreAndFetchFromIndexingNodeAfterPrewarming$35(SharedBlobCacheWarmingServiceIT.java:1263)
```

This assertion is failing on a listener attached to a `ObservableSharedBlobCacheWarmingService` instance, which is a subclass of `SharedBlobCacheWarmingService` that is created in this test and wired into the node via the `TestStatelessPlugin`. I suspect that the `AssertionError` is killing the thread which was supposed to invoke some listener, and the test is timing out because it is waiting on that listener being invoked (perhaps at some number of levels of indirection). So I suspect that the real problem is this assertion firing.

I haven't got anywhere yet with figuring out why that assertion is firing.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -378,9 +378,6 @@ tests:
 - class: org.elasticsearch.xpack.stateless.allocation.EstimatedHeapUsageAllocationDeciderIT
   method: testEstimatedHeapHighWatermarkMonitorTriggersReroute
   issue: https://github.com/elastic/elasticsearch/issues/148246
-- class: org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT
-  method: testCacheIsWarmedBeforeSearchShardRecovery
-  issue: https://github.com/elastic/elasticsearch/issues/148275
 - class: org.elasticsearch.xpack.stateless.StatelessHollowIndexShardsIT
   method: testStressWithoutFailures
   issue: https://github.com/elastic/elasticsearch/issues/148310

--- a/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingServiceIT.java
+++ b/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingServiceIT.java
@@ -518,8 +518,12 @@ public class SharedBlobCacheWarmingServiceIT extends AbstractStatelessPluginInte
         startIndexNode(nodeSettings);
 
         var searchNodeA = startSearchNode(nodeSettings);
-        ensureStableCluster(3);
+        var searchNodeB = startSearchNode(nodeSettings);
+        ensureStableCluster(4);
 
+        // Pin the search shard to searchNodeA so nothing relocates it to searchNodeB
+        // before we install the warming-completed listener on B in phase B. The exclusion is flipped from B to A
+        // at the start of phase B to drive the relocation under test.
         final String indexName = randomIdentifier();
         assertAcked(
             prepareCreate(
@@ -529,6 +533,7 @@ public class SharedBlobCacheWarmingServiceIT extends AbstractStatelessPluginInte
                     .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
                     .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
                     .put(EngineConfig.USE_COMPOUND_FILE, randomBoolean())
+                    .put("index.routing.allocation.exclude._name", searchNodeB)
             )
         );
 
@@ -553,6 +558,11 @@ public class SharedBlobCacheWarmingServiceIT extends AbstractStatelessPluginInte
 
         ensureGreen(indexName);
 
+        // Force the search-recovery warming on searchNodeA to await completion. Without this,
+        // warmCacheForSearchShardRecovery is fire-and-forget on the new-shard-copy path (searchRecoveryTimeout
+        // returns skip()), letting concurrent reads claim cache gaps that warming then skips and the post-warming
+        // ban then trips. See ObservableSharedBlobCacheWarmingService Javadoc for the wrapper semantics.
+        getSharedBlobCacheWarmingService(searchNodeA).setAwaitWarmingForSearchRecovery(true);
         // Verify that we performed pre-warming and don't need to hit the object store on searches
         failObjectStoreAndFetchFromIndexingNodeAfterPrewarming(indexName, searchNodeA, Type.SEARCH);
 
@@ -564,12 +574,14 @@ public class SharedBlobCacheWarmingServiceIT extends AbstractStatelessPluginInte
         disableTransportBlocking(searchNodeA);
         ensureSearchHits(indexName, totalDocs);
 
-        var searchNodeB = startSearchNode(nodeSettings);
-        ensureStableCluster(4);
-
+        // Same await-warming rationale as above: ensure the warming-complete listener (which installs the object-store ban) fires
+        // before the engine opens, so no reader can race the warming task for cache gaps.
+        getSharedBlobCacheWarmingService(searchNodeB).setAwaitWarmingForSearchRecovery(true);
         // The cache also gets pre-warmed when a shard gets relocated to a new node
         failObjectStoreAndFetchFromIndexingNodeAfterPrewarming(indexName, searchNodeB, Type.SEARCH);
-        shutdownNode(searchNodeA);
+        // Flip the exclusion from searchNodeB to searchNodeA so the search shard relocates to searchNodeB.
+        // Simpler than SIGTERM-driven shutdown and equivalent for the recovery path under test.
+        updateIndexSettings(Settings.builder().put("index.routing.allocation.exclude._name", searchNodeA), indexName);
         ensureGreen(indexName);
         assertThat(findSearchShard(resolveIndex(indexName), 0).routingEntry().currentNodeId(), equalTo(getNodeId(searchNodeB)));
         assertHitCount(prepareSearch(indexName).setSize(0), totalDocs);
@@ -1217,32 +1229,6 @@ public class SharedBlobCacheWarmingServiceIT extends AbstractStatelessPluginInte
         };
     }
 
-    private static void shutdownNode(String indexNode) {
-        var shutdownNodeId = client().admin()
-            .cluster()
-            .prepareState(TEST_REQUEST_TIMEOUT)
-            .get()
-            .getState()
-            .nodes()
-            .resolveNode(indexNode)
-            .getId();
-        assertAcked(
-            client().execute(
-                PutShutdownNodeAction.INSTANCE,
-                new PutShutdownNodeAction.Request(
-                    TEST_REQUEST_TIMEOUT,
-                    TEST_REQUEST_TIMEOUT,
-                    shutdownNodeId,
-                    SingleNodeShutdownMetadata.Type.SIGTERM,
-                    "Shutdown for cache warming test",
-                    null,
-                    null,
-                    TimeValue.timeValueMinutes(randomIntBetween(1, 5))
-                )
-            )
-        );
-    }
-
     private void blockAccessToGenerationBeforePostHandoffPreWarmingStarts(String node, long generationToBlock) {
         getSharedBlobCacheWarmingService(node).addBeforeWarmingStartsListener(warmingType -> {
             logger.info("Disabling object store access to generation {}", generationToBlock);
@@ -1384,10 +1370,16 @@ public class SharedBlobCacheWarmingServiceIT extends AbstractStatelessPluginInte
      * {@code warmCacheForSearchShardRecovery}) are not converted into blocking calls. Tests that need "warming done" as a
      * precondition must explicitly await via {@link #addWarmingCompletedListener} or {@link #addMergeWarmingCompletedListener}.
      *
-     * <p>Specific recovery paths can be opted into blocking semantics via {@link #setAwaitWarmingForSearchRecovery(boolean)}
-     * and {@link #setAwaitWarmingForIndexingRecovery(boolean)}. When enabled, the corresponding {@code warmCacheFor*Recovery*}
-     * override blocks the caller thread until the wrapped warming chain (including any registered warming-completed listeners)
-     * completes; this is used by tests that need to install an object-store ban before recovery proceeds.
+     * <p>Specific recovery paths can be opted into await-warming semantics via {@link #setAwaitWarmingForSearchRecovery(boolean)}
+     * and {@link #setAwaitWarmingForIndexingRecovery(boolean)}. Both ensure that any registered {@code warmingCompletedListeners}
+     * fire before recovery resumes, which lets tests install an object-store ban before the engine opens. They differ in how:
+     * <ul>
+     *   <li>{@link #warmCacheForShardRecoveryOrUnhollowing} (indexing) blocks the calling thread until warming completes, because
+     *   in production the recovery thread immediately reads {@code segments_N} after kicking off fire-and-forget warming.</li>
+     *   <li>{@link #warmCacheForSearchShardRecovery} (search) does not block the caller. It routes warming through
+     *   {@link #warmCache} so {@code resumeRecoveryListener} is invoked only after warming (and any
+     *   {@code warmingCompletedListeners}) have completed, deferring the resume step rather than blocking the caller.</li>
+     * </ul>
      */
     private static class ObservableSharedBlobCacheWarmingService extends SharedBlobCacheWarmingService {
 

